unyt/dimensions.py

```diff
@@ -193,3 +196,182 @@
+
+            Check units of `args` and `kwargs`, then run original function.
+
+            Raises
+            ------
+            TypeError
+                If the units do not match.
+
+            """
+            for arg_name, arg_value in chain(zip(names_of_args, args), kwargs.items()):
+                if arg_name in arg_units:  # function argument needs to be checked
+                    dimension = arg_units[arg_name]
+                    if not _has_units(arg_value, dimension):
+                        raise TypeError(
+                            "arg '%s=%s' does not match %s"
+                            % (arg_name, arg_value, dimension)
+                        )
+            return f(*args, **kwargs)
+
+        return new_f
+
+    return check_accepts
+
+
+def returns(r_unit):
+    """Decorator for checking function return units.
+
+    Parameters
+    ----------
+    r_unit: :py:class:`sympy.core.symbol.Symbol`
+        SI base unit (or combination of units), eg. length/time
+        of the value returned by the original function
+
+    Examples
+    --------
+    >>> import unyt as u
+    >>> from unyt.dimensions import length, time
+    >>> @returns(length)
+    ... def f(a, v):
+    ...     return a * v
+    ...
+    >>> res = f(a= 2 * u.s, v = 3 * u.m/u.s)
+    >>> print(res)
+    6 m
+    >>> @returns(length/time)
+    ... def f(a, v):
+    ...     return a * v
+    ...
+    >>> f(a= 2 * u.s, v = 3 * u.m/u.s)
+    Traceback (most recent call last):
+    ...
+    TypeError: result '6 m' does not match (length)/(time)
+
+    """
+
+    def check_returns(f):
+        """Decorates original function.
+
+        Parameters
+        ----------
+        f : function
+            Function being decorated.
+
+        Returns
+        -------
+        new_f: function
+            Decorated function.
+
+        """
+
+        @wraps(f)
+        def new_f(*args, **kwargs):
+            """The decorated function, which checks the return unit.
+
+            Raises
+            ------
+            TypeError
+                If the units do not match.
+
+            """
+            result = f(*args, **kwargs)
+            if not _has_units(result, r_unit):
+                raise TypeError("result '%s' does not match %s" % (result, r_unit))
+            return result
+
+        return new_f
+
+    return check_returns
+
+
+def _has_units(quant, dim):
+    """Checks the argument has the right dimensionality.
+
+    Parameters
+    ----------
+    quant : :py:class:`unyt.array.unyt_quantity`
+        Quantity whose dimensionality we want to check.
+    dim : :py:class:`sympy.core.symbol.Symbol`
+        SI base unit (or combination of units), eg. length/time
+
+    Returns
+    -------
+    bool
+        True if check successful.
+
+    Examples
+    --------
+    >>> import unyt as u
+    >>> from unyt.dimensions import length, time
+    >>> _has_units(3 * u.m/u.s, length/time)
+    True
+    >>> _has_units(3, length)
+    False
```

Comment:
Is it bad practice or just unnecessary work? :)

Reply:
Just unnecessary work. It helps of course for developers to see the documentation but since private functions don't show up in the API docs it's not user-facing.